### PR TITLE
[syntax] Allow unicode identifiers

### PR DESCRIPTION
**Pull request by artagnon**
~~Depends on #89.~~

### CODE REVIEW

**Comment by artagnon** [comment]
I just realized that this patch doesn't depend on the modernize branch, and have rebased onto master.

File: client/syntaxes/coq.tmLanguage [hunk 1]
```diff
@@ -36,7 +36,7 @@
 
         <dict>
             <key>match</key>
-            <string>\b(Require|Import|Export|Local|Global|Open|Scope)\b</string>
+            <string>\b(Require|Import|Export|Local|Global|Open|Scope|Context)\b</string>
```
Comment (maximedenes) [inline client/syntaxes/coq.tmLanguage]: Why add `Context` here? TBH, I have no idea what this list was about in the first place.

File: client/syntaxes/coq.tmLanguage [hunk 2]
```diff
@@ -65,7 +65,7 @@
 
         <dict>
             <key>match</key>
-            <string>(Program Definition|Program Fixpoint|Program CoFixpoint|Program Function|Function|CoFixpoint|Fixpoint|Definition|Let)\s+([a-zA-Z_][a-zA-Z0-9_']*)</string>
+            <string>(Program Definition|Program Fixpoint|Program CoFixpoint|Program Function|Function|CoFixpoint|Fixpoint|Definition|Let|Section|Arguments)\s+(([\p{L}-]|[0-9]|[_'])+)</string>
```
Comment (maximedenes) [inline client/syntaxes/coq.tmLanguage]: Why add `Section` and `Arguments` here? They are not definition-like commands.

File: client/syntaxes/coq.tmLanguage [hunk 3]
```diff
@@ -125,7 +125,7 @@
 
         <dict>
             <key>match</key>
-            <string>\b(Hint|Constructors|Resolve|Rewrite|Proof|Ltac|Qed|Defined|Implicit|Set|Arguments|Tactic Notation|Notation|Section|Module Type|Module|End|Variables|Record|Check|Print|Eval)\b</string>
+            <string>(Hint|Constructors|Resolve|Rewrite|Proof|Ltac|Qed|Defined|Implicit|Set|Tactic Notation|Notation|Reserved Notation|Module Type|Module|End|Variables|Check|Print|Eval|Next Obligation)</string>
```
Comment (maximedenes) [inline client/syntaxes/coq.tmLanguage]: I see you change the list here too, but again, I'm not sure what's the idea behind this list. They are not keywords AFAIK, or at least not all of them.

File: client/syntaxes/coq.tmLanguage [hunk 4]
```diff
@@ -45,7 +45,7 @@
 
         <dict>
             <key>match</key>
-            <string>(Lemma|Theorem|Goal|Example)\s+([a-zA-Z][a-zA-Z0-9_']*)</string>
+            <string>(Lemma|Theorem|Goal|Example)\s+(([\p{L}-]|[0-9]|[_'])+)</string>
```
Comment (maximedenes) [inline client/syntaxes/coq.tmLanguage]: Doesn't this allow identifiers beginning with numbers?

Comment (artagnon) [reply client/syntaxes/coq.tmLanguage]: Good catch; will fix shortly.

**Comment by artagnon** [comment]
Better?

File: client/syntaxes/coq.tmLanguage [hunk 5]
```diff
@@ -45,7 +45,7 @@
 
         <dict>
             <key>match</key>
-            <string>(Lemma|Theorem|Goal|Example)\s+([a-zA-Z][a-zA-Z0-9_']*)</string>
+            <string>(Lemma|Theorem|Goal|Example)\s+([\p{L}-]([\p{L}-]|[0-9]|[_'])+)</string>
```
Comment (maximedenes) [inline client/syntaxes/coq.tmLanguage]: I'm not very familiar with vscode's regexps, what does the first hyphen mean in the pattern?

Comment (artagnon) [reply client/syntaxes/coq.tmLanguage]: `\p{L}` matches `à`: I've just left out the end of the range, which would be the last unicode character. It's not a vscode-specific regex: you can look up any unicode regex [reference sheet](https://www.regular-expressions.info/refunicode.html).

Comment (maximedenes) [reply client/syntaxes/coq.tmLanguage]: But you want to match only letters, don't you?

Comment (maximedenes) [reply client/syntaxes/coq.tmLanguage]: More concretely, I'm wondering why `[\p{L}-]` rather than `\p{L}`.

Comment (artagnon) [reply client/syntaxes/coq.tmLanguage]: `θIO` is a valid identifier: getting it to highlight correctly was the main motivation behind this patch.

Comment (artagnon) [reply client/syntaxes/coq.tmLanguage]: Think of `[\p{L}-]` as the unicode equivalent of `[a-]` (or `[a-z]`). You don't want every identifier to begin with an `à`, right?

Comment (maximedenes) [reply client/syntaxes/coq.tmLanguage]: Doesn't `b` also match `\p{L}`?

Comment (maximedenes) [reply client/syntaxes/coq.tmLanguage]: Your documentation points to https://www.regular-expressions.info/unicode.html#category, which says: `\p{L} or \p{Letter}: any kind of letter from any language`

Comment (artagnon) [reply client/syntaxes/coq.tmLanguage]: Thanks for catching this: I read the documentation in a hurry; fixed now.

File: client/syntaxes/coq.tmLanguage [hunk 6]
```diff
@@ -45,7 +45,7 @@
 
         <dict>
             <key>match</key>
-            <string>(Lemma|Theorem|Goal|Example)\s+([a-zA-Z][a-zA-Z0-9_']*)</string>
+            <string>(Lemma|Theorem|Goal|Example)\s+(\p{L}(\p{L}|[0-9]|[_'])+)</string>
```
Comment (maximedenes) [inline client/syntaxes/coq.tmLanguage]: And why do you exclude one-letter identifiers?

Comment (artagnon) [reply client/syntaxes/coq.tmLanguage]: Thanks again; fixed.

Comment (maximedenes) [reply client/syntaxes/coq.tmLanguage]: Also, you can keep `[0-9]` and `[_']` merged, as in the original code. (sorry for the multiple comments, I just noticed this)

**Comment by artagnon** [comment]
I don't have write access, so could you merge this?

**Pull request merged**